packages/pymedphys_deliverydata/src/pymedphys_deliverydata/object/exportable.py

```diff
@@ -0,0 +1,124 @@
+from pymedphys_mudensity.mudensity import calc_mu_density
+
+from ..base import DeliveryDataBase
+from ..dicom import (
+    delivery_data_to_dicom,
+    dicom_to_delivery_data,
+    gantry_tol_from_gantry_angles)
+from ..utilities import (
+    filter_out_irrelevant_control_points,
+    get_all_masked_delivery_data,
+    get_metersets_from_delivery_data)
+
+
+class DeliveryData(DeliveryDataBase):
+    def __new__(cls, *args):
+        new_args = (
+            to_tuple(arg)
+            for arg in args
+        )
+        return super().__new__(cls, *new_args)
+
+    @classmethod
+    def from_delivery_data_base(cls, delivery_data_base):
+        if type(delivery_data_base) is type(cls):
+            return delivery_data_base
+
+        return cls(*delivery_data_base)
+
+    @classmethod
+    def from_dicom(cls, dataset):
+        return cls.from_delivery_data_base(
+            dicom_to_delivery_data(dataset))
+
+    @classmethod
+    def empty(cls):
+        return cls(
+            tuple(),
+            tuple(),
+            tuple(),
+            tuple(
+                tuple(tuple(), tuple())
+            ),
+            tuple(
+                (tuple(), tuple())))
+
+    def to_dicom(self, template):
+        return delivery_data_to_dicom(self, template)
+
+    @functools.lru_cache()
+    def filter_cps(self):
+        return filter_out_irrelevant_control_points(self)
+
+    @functools.lru_cache()
+    def mask_by_gantry(self, angles: Union[tuple, float, int], tolerance=3):
+        iterable_angles: tuple
+
+        try:
+            _ = iter(angles)  # type: ignore
+            iterable_angles = tuple(angles)  # type: ignore
+        except TypeError:
+            # Not iterable, assume just one angle provided
+            iterable_angles = (angles,)
+
+        return get_all_masked_delivery_data(self, iterable_angles, tolerance)
+
+    @functools.lru_cache()
+    def metersets(self, gantry_angles, gantry_tolerance):
+        self.mask_by_gantry(gantry_angles, gantry_tolerance)
+        return get_metersets_from_delivery_data(
+            self.mask_by_gantry(gantry_angles, gantry_tolerance))
+
+    def mudensity(self, gantry_angles=None, gantry_tolerance=None, grid_resolution=1):
+        if gantry_angles is None:
+            gantry_angles = 0
+            gantry_tolerance = 500
+        elif gantry_tolerance is None:
+            gantry_tolerance = gantry_tol_from_gantry_angles(gantry_angles)
+
+        masked_by_gantry = self.mask_by_gantry(gantry_angles, gantry_tolerance)
+
+        mudensities = []
+        for delivery_data in masked_by_gantry:
+            mudensities.append(calc_mu_density(
+                delivery_data.monitor_units,
+                delivery_data.mlc,
+                delivery_data.jaw,
+                grid_resolution=grid_resolution))
+
+        # TODO: This is worth discussion whether or not we should do this
+        if len(mudensities) == 1:
+            return mudensities[0]
```

Comment:
If one or no gantry angle is provided mudensity is returned as a numpy array.

If more than one gantry angle is provided mudensity is returned as a list of numpy arrays.

Do you think this is a good idea?

Reply:
I think I'd slightly prefer the type to be consistent, even if it does mean that the user has to go through the tedium of `mudensities[0]`

Also, imagine some sort of batch processing that loops through the returned `mudensities`. You don't want code intended to loop through a list of mudensities (that theoretically could support a list with a single item) to instead loop through the numpy array...

Reply:
Yup, that's what I was thinking. As I wrote it I went ... hmmm.... not sure.